### DISCUSSION TITLE
Does Langium a good fit for a Svelte-like language?

### DISCUSSION THREAD

**Discussion by arthur-fontaine** [eclipse-langium/langium#859] [Q&A] [answered]
Hello,

I need to implement a syntax that looks like that:
```
<script>
  import { createState, createWatcher, createComputed } from 'kazam'
  import { Text } from './Text'

  const [counter, setCounter] = createState(0)
  const doubleCounter = createComputed(() => counter * 2)

  const [colors, setColors] = createState(['red', 'orange', 'yellow'])

  const [fruits, setFruits] = createState({ apple: 1, banana: 2, pear: 3 })

  const [cats, setCats] = createState([{ name: 'Fluffy' }, { name: 'Whiskers' }])

  createWatcher((counter) => console.log(counter), counter)
</script>

@if (counter % 2 === 0) {
  <p>Counter is even</p>
} @else if (counter % 2 !== 0) {
  <p>Counter is odd</p>
} @else {
  <p>Counter is not a number</p>
}

<p class="a-class" on:click={console.log}>Counter: {counter}</p>
@for (i = 0; i < counter; i++) {
  <p>i: {i}</p>
}

@for (const color of colors) {
  <p>{color}</p>
}

@for (const fruit in fruits) {
  <p>{fruit}: {fruits[fruit]}</p>
}

@each (cats as cat) {
  <p>{cat.name}</p>
}
```

I just learned about Langium. I think it would be possible to use recursive rules, create `terminal` for `<startTag>` and `</endTag>` and create a validation function that check if they are the same tag...
But I don't know if Langium really fits for my needs. While I was trying to implement the grammar, I had a lot of `"Token: ->}<- can never be matched."`-like errors ([playground link](https://langium.org/playground?grammar=OYJwhgthYgBA0mAXgKBQUwHYBcQE9YBZAewBN0AbALhVjtgApsxgBnAagF4AVFgSgBUAbjQAzcMAhZssAMrMQ2XsBr1YAcgA862KwVKWAOUjpOASQAijMNlwBLAEYBXbOg6cAgrZCOX6QSIospSiAMIUxKx2mMDKqvTyMAbAGgD0AHzqgYbEmMEUYRFRMXG0CfrKGpmMAMYAFnYUpCBYXMqCGtppOlikysZS5lbqmYHc6AAe2PF0rlM8AKIAGtxjLDNyIeGR0bEssAA%2BsDl5W0W7lUcAMsTAdjWHsONTjwsTAA4trFG5gV72zlcGwYdlYCwAbtIAPycdS5KjqPhQ2CYExDDSw2DgsAUJymBiybgAJTMhgA4q8Pl8fpg%2BIE3p83DSNuoAN49KlMuy5dHqAC%2BWTQNzuNQ2dlInCuAHkyWZQgB9SyMd4wEyuECSmVy%2BUABQ8RI8hAW3AWRKRGnZtQaTRamDa-AEGgFgXVEGiONgliosFSAG15WAALRIDyBgBaAF1fQAdADu8ojAlSIld7oocmJpLJ3tSACJfQA9XOJ3MHdSF9SJ9TJtCp1Hp6WyhV6g1Gk1EnPRhiFvgR9jRvg1usexvar0%2BgAC-qDIfDUbj8tgiZj8YjNZQw-TJpWOcLmgnffXDVI5EwsE3sAA6rJOxwa0A&content=DwBwBAxgNghgznAvAIgLYFoAMywHsB2AXNAJYQDWiA3gC4CmcNAvgHwBQYYwJqA5mHABOEFAAsaNEHEIB6GSQB0PXgFdBCiLlQyAjAC0AigGY9AdnUh8vHDCg0UASVQxedHDPadgMMKMF0AMzEJKVkZAHdIhQBPXBUaFQAjOg0tCJgaCFEAfgA3RH1jM0EcGhhBV3tkAH1E2HxyZBZkcUlpOUjwmLiE5NTtcIysvILDE3NkYBkYdimQdjYAAToYLLAACggMuDB4SAyASjAqDi55qi3mOfYmNiA)).

My question: **Is Langium suitable for creating a language with syntax similar to Svelte / HTML**, and are there any examples available?

Thanks in advance!

**Comment by msujew** [comment, marked as answer]
Hey @arthur-fontaine, thanks for your question!

I've updated your playground to work as expected, see [here](https://langium.org/playground?grammar=OYJwhgthYgBA0mAXgKBQUwHYBcQE9YBZAewBN0AbALhVjtgApsxgBnAagF4AVFgSgBUAbjQAJbBAq9gNerADkAHnmxWzENmkA5SOk4BJACKMw2XAEsARgFds6DpwCCZkFdvpBjVpQBmAYQpiVnNMYAB%2BTnkAegA%2BFQAfBTjGAGMAC3MKUhAsLmlPJRVolSxSbV0DY3k4vhEUbnQAD2xZejtmzgBlbgAlfS0AcTrpVrpxSWlYRIAZYmBzFKnYBualgFFGgAcc1mDiTDrnCxs7UcZzVjWANyxsCPl9qnk%2BMNhMCqMFSNgrsAprPQMbp9QbrLY7PaYWpoDbbeyQs7yADeJXB8PM%2B0qCgAvvI6rN5ikzuZSJxpgB5Ab6PwAfU%2BDE2MF0dhAZMp1JpAAVHD1HIQ1tw1j0XgoUakMlkcpg8vwBDi8WgWRAQn9YEYqLAogBtGlgAC0SEceoAWgBdLUAHQA7jTTQIoiIlSqKLBgf0BhqogAiLUAPS9dq98XkfvkduidSd7xdFKptO5vP5gp6notDD9fFN7AtfAdKCjqtjHPVmoAAjr9YaTebrTTYHbLTbTXmUBlSORMLACy6AOqdVMcPNAA&content=DwBwBAxgNghgznAvAIgLYFoAMywHsB2AXNAJYQDWiA3gC4CmcNAvgHwBQYYwJqA5mHABOEFAAsaNEHEIB6GSQB0PXgFdBCiLlQyAjAC0AigGY9AdnUh8vHDCg0UASVQxedHDPadgMMKMF0AMzEJKVkZAHdIhQBPXBUaFQAjOg0tCJgaCFEAfgA3RH1jM0EcGhhBV3tkAH1E2HxyZBZkcUlpOUjwmLiE5NTtcIysvILDE3NkYBkYdimQdjYAAToYLLAACggMuDB4SAyASjAqDi55qi3mOfYmNiA). The `TEXT` terminal messes with a lot of stuff in Langium, due to it being effectively being a catch-all terminal. We perform a few heuristics to optimize lexing/parsing performance, including reordering of token types. In this case it gets moved in front of all other tokens due to it's ability to match a whitespace character, which prevents everything else from being lexed. So in a real langium project you might need to fine-tune the token generation.

There was also another issue related to ambiguous grammar rules, but that is easily fixed.

Anyway, HTML should be fairly straightforward to build. Note that Langium is best suited for languages with static typing. I see a lot of `ID`s in places where cross references should be used, since that's the real benefit that Langium provides you (i.e. cross reference resolution + all the language services, go-to definitions, finding references, etc.). As long as your language somehow fits in the area of static typing + parseable by an LL parser, Langium is usually a good fit.

Also, if you're looking for a more in-depth example on how to build an expression language, take a look at the [langium-lox](https://github.com/langium/langium-lox) repo.

**Reply by arthur-fontaine** [reply]
Hi, thank you very much for your comprehensive answer! I'll look into it.

Wouldn't it be a good idea to add an **Examples** section in the README mentioning the [langium-lox](https://github.com/langium/langium-lox) repo? I think it would be nice to add the **Examples** section to the navigation sidebar in the documentation too.

**Reply by trusktr** [reply]
The more examples the better. Yes please.

**Reply by trusktr** [reply]
> Langium is best suited for languages with static typing

How would we, for example, begin to make a language like lox and type check that a variable that holds a string is not assigned to a variable that holds numbers?

**Reply by trusktr** [reply]
Ah, https://github.com/langium/langium/discussions/706 points to how to implement type checking (and nice hints at the future of a framework for easily defining an language's type system).

**Reply by msujew** [reply]
> How would we, for example, begin to make a language like lox and type check that a variable that holds a string is not assigned to a variable that holds numbers?

Note that the langium-lox implementation is a statically typed dialect of the original Lox language. Dealing with dynamic/runtime typing in Langium is a pain and not on our happy-path. As for the assignability question: We implement a [validation for binary operation](https://github.com/langium/langium-lox/blob/f28627cbe0ce9ce2d714a96a04f5ee3c95c53d14/src/language-server/lox-validator.ts#L87-L93) (one of which is the `=` operator) and see whether the operation is legal for the inferred type of the left side and the inferred type of the right side.
